### PR TITLE
[CPU] remove redundant reorders

### PR DESCRIPTION
**Pull request by mateusztabaka**
For cases when node has multiple edges pointing to (equivalent) reorders,
we can replace that with a single reorder, then update its child edges accordingly.
For example, below graph:
```
              +---------+
              |  node1  |
              +---------+
                 |   |
          -------    -------
          |                |
          v                v
    +-----------+    +-----------+
    |  reorder  |    |  reorder  |
    +-----------+    +-----------+
          |                |
          v                v
     +---------+      +---------+
     |  node2  |      |  node3  |
     +---------+      +---------+

```
gets converted to:
```
              +---------+
              |  node1  |
              +---------+
                   |
                   v
             +-----------+
             |  reorder  |
             +-----------+
                 |   |
          -------    -------
          |                |
          v                v
     +---------+      +---------+
     |  node2  |      |  node3  |
     +---------+      +---------+
```

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 1]
```diff
@@ -1523,12 +1523,16 @@ void MKLDNNGraphOptimizer::FuseEltwiseAndSimple(MKLDNNGraph &graph) {
     }
 }
 
-void MKLDNNGraphOptimizer::DropDoubleReorders(MKLDNNGraph &graph) {
+void MKLDNNGraphOptimizer::DropRedundantReorders(MKLDNNGraph &graph) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1526]: I am not sure this is a good idea to extend existing DropDoubleReorders from code readability standpoint. At the same time it allows us to have only one pass through all the graph nodes which better for FIL metric. Maybe it is a good idea to extract different branches into separate functions with meaningful names (like VerticalReorderFusion and HorizontalReorderFusion)

**Comment by dmitry-gorokhov** [comment]
@antonvor please review the PR

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 1 above)
Comment (mateusztabaka) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:1526]: > Maybe it is a good idea to extract different branches into separate functions with meaningful names (like VerticalReorderFusion and HorizontalReorderFusion)

Sounds good to me. The function has gotten too long indeed.

**Review by antonvor — COMMENTED** [review]
LGTM, agree with the comment above. Small question: should we update the reorder node name? Now the reorder node can have several outputs, maybe we should add all the names of the output nodes as a suffix?

**Comment by mateusztabaka** [comment]
> LGTM, agree with the comment above. Small question: should we update the reorder node name? Now the reorder node can have several outputs, maybe we should add all the names of the output nodes as a suffix?

I was thinking of that, but I noticed that reorder node's names tend to be long already so I was kind of reluctant to extend reorder node name because of that fact

**Comment by antonvor** [comment]
@mateusztabaka maybe we can add test that checks horizontal fusing transformation?

File: src/plugins/intel_cpu/src/mkldnn_graph_optimizer.cpp [hunk 2]
```diff
@@ -1562,44 +1562,150 @@ void MKLDNNGraphOptimizer::FuseEltwiseAndSimple(MKLDNNGraph &graph) {
...
+    // Map value is a list of node edges indices that point to the redundant reorders
+    std::unordered_map<std::string, std::vector<size_t>> reorders;
+    auto childEdges = node->getChildEdges();
+    for (size_t edgeId = 0; edgeId < childEdges.size(); edgeId++) {
+        const auto edge = childEdges[edgeId].lock();
+        if (edge->getChild()->getType() != Reorder)
+            continue;
+        const auto reorder = edge->getChild();
+        const auto reorderPtr = dynamic_cast<MKLDNNReorderNode*>(reorder.get());
+        if (!reorderPtr || !reorderPtr->getIsOptimized())
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/plugins/intel_cpu/src/mkldnn_graph_optimizer.cpp]: You're right. I fixed this in the latest commit.

File: src/tests/functional/plugin/cpu/subgraph_tests/src/redundant_reorders_test.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            } else if (layer_type == "Concat") {
+                for (size_t i = 0; i < n->get_input_size(); i++)
+                    ASSERT_NE(reorder, n->get_input_node_shared_ptr(i));
+            }
+        }
+    }
+};
+
+TEST_F(RedundantReordersTest, smoke_RedundantReordersTest) {
+    Run();
```
Comment (antonvor) [inline src/tests/functional/plugin/cpu/subgraph_tests/src/redundant_reorders_test.cpp:62]: small: can you add `SKIP_IF_CURRENT_TEST_IS_DISABLED()` here?

**Review by antonvor — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@mateusztabaka could you please rebase the PR on latest master

**Pull request closed without merging**